relay-profiling/src/android.rs

```diff
@@ -126,97 +126,46 @@ impl AndroidProfile {
     }
 }
 
-pub fn expand_android_profile(payload: &[u8]) -> Result<Vec<Vec<u8>>, ProfileError> {
-    let profile = parse_android_profile(payload)?;
-    let mut items: Vec<Vec<u8>> = Vec::new();
-
-    if profile.transactions.is_empty() && profile.has_transaction_metadata() {
-        match serde_json::to_vec(&profile) {
-            Ok(payload) => items.push(payload),
-            Err(_) => {
-                return Err(ProfileError::CannotSerializePayload);
-            }
-        };
+fn parse_profile(payload: &[u8]) -> Result<AndroidProfile, ProfileError> {
+    let mut profile: AndroidProfile =
+        serde_json::from_slice(payload).map_err(ProfileError::InvalidJson)?;
 
-        return Ok(items);
+    if profile.transactions.is_empty() && !profile.has_transaction_metadata() {
+        return Err(ProfileError::NoTransactionAssociated);
     }
 
-    for transaction in &profile.transactions {
-        let mut new_profile = profile.clone();
-
-        new_profile.profile_id = EventId::new();
-        new_profile.set_transaction(transaction);
-        new_profile.transactions.clear();
-
-        let clock = new_profile.profile.clock;
-        let start_time = new_profile.profile.start_time;
-
-        new_profile.profile.events.retain_mut(|event| {
-            let event_timestamp = get_timestamp(clock, start_time, event.time);
-            if transaction.relative_start_ns <= event_timestamp
-                && event_timestamp <= transaction.relative_end_ns
-            {
-                let relative_start = match clock {
-                    Clock::Cpu => Duration::from_millis(transaction.relative_cpu_start_ms),
-                    Clock::Wall | Clock::Dual | Clock::Global => {
-                        Duration::from_nanos(transaction.relative_start_ns)
-                    }
-                };
-                event.time = substract_to_timestamp(event.time, relative_start);
-                true
-            } else {
-                false
-            }
-        });
-
-        if new_profile.profile.events.is_empty() {
-            continue;
+    if let Some(transaction) = profile.transactions.drain(..).next() {
+        if !transaction.valid() {
+            return Err(ProfileError::InvalidTransactionMetadata);
         }
 
-        match serde_json::to_vec(&new_profile) {
-            Ok(payload) => items.push(payload),
-            Err(_) => {
-                return Err(ProfileError::CannotSerializePayload);
-            }
-        };
+        profile.transaction = Some(transaction.clone());
+
+        // this is for compatibility
+        profile.active_thread_id = transaction.active_thread_id;
+        profile.duration_ns = transaction.duration_ns();
+        profile.trace_id = transaction.trace_id;
+        profile.transaction_id = transaction.id;
+        profile.transaction_name = transaction.name;
+    }
+
+    if !profile.sampled_profile.is_empty() {
+        profile.parse()?;
```

Comment:
Does this `parse()` and subsequent removal of events have to happen after setting the transaction? If not, I would turn this around to avoid a possibly redundant operation.

Reply:
This is just the heaviest operation you can think of in terms of parsing compared to just validating a few things first.